jdaviz/configs/mosviz/plugins/parsers.py

```diff
@@ -645,66 +691,73 @@ def mos_niriss_parser(app, data_dir, obs_label=""):
-
                         spec2d.meta['INSTRUME'] = 'NIRISS'
                         spec2d.meta['mosviz_row'] = len(spec_labels_2d)
 
-                        label = f"{filter_name} Source {temp[sci].header['SOURCEID']} spec2d {orientation}"  # noqa
-                        ra, dec = pupil_id_dict[filter_name][temp[sci].header["SOURCEID"]]
+                        label = (f"{filter_name} Source "
+                                 f"{temp[sci].header['SOURCEID']} spec2d "
+                                 f"{orientation}")  # noqa
+                        add_to_glue[label] = spec2d
+
+                        # update labels for table viewer
+                        ra, dec = cat_id_dict[temp[sci].header["SOURCEID"]]
                         ras.append(ra)
                         decs.append(dec)
                         image_add.append(image_dict[filter_name])
                         spec_labels_2d.append(label)
-
-                        add_to_glue[label] = spec2d
-
                         filters.append(filter)
 
-    spec_labels_1d = []
-    for f in ["1D Spectra C", "1D Spectra R"]:
+    # Parse 1D spectra
+    file_labels_1d = [k for k in file_lists.keys() if k.startswith("1D")]
 
-        for fname in file_lists[f]:
-            print(f"Loading: {f} sources")
+    for flabel in file_labels_1d:
+        for fname in file_lists[flabel]:
+            print(f"Loading: {flabel} sources")
 
             with fits.open(fname, memmap=False) as temp:
+                # treat all HDUs missing SRCTYPE attribute as extended sources
                 # TODO: Remove this once valid SRCTYPE values are present in all headers
                 for hdu in temp:
-                    if ("SRCTYPE" in hdu.header and
-                            (hdu.header["SRCTYPE"] in ("POINT", "EXTENDED"))):
+                    if ("SRCTYPE" in hdu.header
+                            and (hdu.header["SRCTYPE"] in ("POINT", "EXTENDED"))):
                         pass
                     else:
                         hdu.header["SRCTYPE"] = "EXTENDED"
 
+                # read all HDUs with SpectrumList, then only keep those that
+                # correspond with sources in catalog
+                # (read() is slow... would also LOVE to do this in one step!)
                 specs = SpectrumList.read(temp, format="JWST x1d multi")
-                filter_name = [x
-                               for x in fname.split("/")[-1].split("_")
-                               if x[0] == "F" or x[0] == "f"][0]
+                specs_cut = [sp for sp in specs if sp.meta['header']['SOURCEID']
+                             in cat_id_dict.keys()]
 
-                # Orientation denoted by "C" or "R"
-                orientation = f[-1]
+                filter_name = fits.getheader(fname, ext=0).get('PUPIL')
 
-                for spec in specs:
-                    # Make metadata layout conform with other viz.
-                    spec.meta = standardize_metadata(spec.meta)
```

Comment:
You might want to keep this line to inherit metadata from FITS header? Maybe this is why your CI is failing? Hard to tell from the diff.

Reply:
This is was indeed the issue. The failing lines pass now for me locally, so let's hope it's reflected in the newest CI results.

Reply:
Yup, tests pass now. 😸